CommandLine/libGMX/gmx.cpp

```diff
@@ -0,0 +1,382 @@
+/** Copyright (C) 2018 Greg Williamson, Robert B. Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <pugixml.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <sstream>
+#include <vector>
+#include <algorithm>
+
+#include "gmx.h"
+
+namespace gmx {
+
+inline std::string string_replace_all(std::string str, std::string substr, std::string nstr)
+{
+  size_t pos = 0;
+  while ((pos = str.find(substr, pos)) != std::string::npos)
+  {
+    str.replace(pos, substr.length(), nstr);
+    pos += nstr.length();
+  }
+  return str;
+}
+
+inline std::string extract_file_path(std::string dir, std::string value) {
+  value = string_replace_all(value, "\\", "/");
+  const std::string end = dir.substr(dir.find_last_of("/") + 1);
+  if (value.find(end) == 0)
```

Comment:
Why are you checking if `value` starts with the last path of `dir`? Are you assuming the next char will be `/`? Or do you expect the two to be equal? This function is doing more than just extracting a path.

Reply:
Well sounds have `<origname>sound\audio\soundname.ext</origname>` and `<data>soundname.ext</data>` whereas sprites and backgrounds have only `<data>images\bkgsprname.png</data>`

So, it should be clear that what I am doing is checking if the end of the full dir is at the beginning of the data/origname fields. Then I know whether to prepend the whole thing or just a part of it.

I was consider replacing `(file_path) = true` custom FieldOption with `(file_path) = "the\prefix\for\this\file\field"`

What do you think about that instead?